File: LeetCode/872/solution.java
// 872. Leaf-Similar Trees
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  public boolean leafSimilar(TreeNode root1, TreeNode root2) {
    List<Integer> list1 = new ArrayList<>();
    List<Integer> list2 = new ArrayList<>();
    
    dfs(root1, list1);
    dfs(root2, list2);

    return list1.equals(list2);
  }

  private void dfs(TreeNode node, List<Integer> list) {
    if (node == null) return;
    
    if (node.left == null && node.right == null){
      list.add(node.val);
      return;
    }

    dfs(node.left, list);
    dfs(node.right, list);
  }
}

File: LeetCode/872/solution.cpp
// 872. Leaf-Similar Trees
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
      vector<int> list1, list2;

      dfs(root1, list1);
      dfs(root2, list2);
      
      return list1 == list2;
    }

    void dfs(TreeNode* node, vector<int>& list) {
      if (!node) return;

      if (!node->left && !node->right) {
        list.push_back(node->val);
        return;
      }

      dfs(node->left, list);
      dfs(node->right, list);
    }
};